internals-js/src/federation.ts

```diff
@@ -463,6 +467,25 @@ export class FederationMetadata {
     return this._sharingPredicate;
   }
 
+  private keysPredicate(): (field: FieldDefinition<CompositeType>) => boolean {
+    if (!this._keysPredicate) {
+      this._keysPredicate = computeKeys(this.schema);
+    }
+    return this._keysPredicate;
+  }
+
+  private fieldUsedPredicate(): (field: FieldDefinition<CompositeType>) => boolean {
+    if (!this._fieldUsedPredicate) {
+      const usedFields = Array.from(collectUsedFields(this));
```

Comment:
I think we should use a `Set` for this. On large schema, there could be quite a few "used" fields overall, and `isUsedField` might also be used for many fields, so the linear scan could realistically add up. It's fine (even admittedly cleaner) that `collectUsedFields` now return `FieldDefinition`s rather than just the coordinates, but I would still convert back to a `Set<string>` at that stage so that `_fieldUsedPredicate` can be a simple set lookup.